# Zad1

In [6]:
import pandas as pd
import numpy as np

data = pd.read_csv(r'C:\Users\THINK\Documents\ML\CW2\phpMYEkMl.csv')
data = data.replace('?',np.nan) 
data["CabinReduced"] =  data["cabin"].astype(str).str[0]
#Cant find better idea 
data["CabinReduced"] = data["CabinReduced"].replace("n",np.nan)

data.head(40)

pclass  survived                                               name  \
0        1         1                      Allen, Miss. Elisabeth Walton   
1        1         1                     Allison, Master. Hudson Trevor   
2        1         0                       Allison, Miss. Helen Loraine   
3        1         0               Allison, Mr. Hudson Joshua Creighton   
4        1         0    Allison, Mrs. Hudson J C (Bessie Waldo Daniels)   
5        1         1                                Anderson, Mr. Harry   
6        1         1                  Andrews, Miss. Kornelia Theodosia   
7        1         0                             Andrews, Mr. Thomas Jr   
8        1         1      Appleton, Mrs. Edward Dale (Charlotte Lamson)   
9        1         0                            Artagaveytia, Mr. Ramon   
10       1         0                             Astor, Col. John Jacob   
11       1         1  Astor, Mrs. John Jacob (Madeleine Talmadge Force)   
12       1         1                      Aubart, Mme. Leontine Pauline   
13       1         1                       Barber, Miss. Ellen 'Nellie'   
14       1         1               Barkworth, Mr. Algernon Henry Wilson   
15       1         0                                Baumann, Mr. John D   
16       1         0                           Baxter, Mr. Quigg Edmond   
17       1         1    Baxter, Mrs. James (Helene DeLaudeniere Chaput)   
18       1         1                              Bazzani, Miss. Albina   
19       1         0                               Beattie, Mr. Thomson   
20       1         1                      Beckwith, Mr. Richard Leonard   
21       1         1   Beckwith, Mrs. Richard Leonard (Sallie Monypeny)   
22       1         1                              Behr, Mr. Karl Howell   
23       1         1                              Bidois, Miss. Rosalie   
24       1         1                                  Bird, Miss. Ellen   
25       1         0                                Birnbaum, Mr. Jakob   
26       1         1                            Bishop, Mr. Dickinson H   
27       1         1            Bishop, Mrs. Dickinson H (Helen Walton)   
28       1         1                             Bissette, Miss. Amelia   
29       1         1          Bjornstrom-Steffansson, Mr. Mauritz Hakan   
30       1         0                       Blackwell, Mr. Stephen Weart   
31       1         1                                   Blank, Mr. Henry   
32       1         1                            Bonnell, Miss. Caroline   
33       1         1                           Bonnell, Miss. Elizabeth   
34       1         0                           Borebank, Mr. John James   
35       1         1                           Bowen, Miss. Grace Scott   
36       1         1                        Bowerman, Miss. Elsie Edith   
37       1         1      Bradley, Mr. George ('George Arthur Brayton')   
38       1         0                            Brady, Mr. John Bertram   
39       1         0                                 Brandeis, Mr. Emil   

       sex     age  sibsp  parch    ticket      fare    cabin embarked boat  \
0   female      29      0      0     24160  211.3375       B5        S    2   
1     male  0.9167      1      2    113781    151.55  C22 C26        S   11   
2   female       2      1      2    113781    151.55  C22 C26        S  NaN   
3     male      30      1      2    113781    151.55  C22 C26        S  NaN   
4   female      25      1      2    113781    151.55  C22 C26        S  NaN   
5     male      48      0      0     19952     26.55      E12        S    3   
6   female      63      1      0     13502   77.9583       D7        S   10   
7     male      39      0      0    112050         0      A36        S  NaN   
8   female      53      2      0     11769   51.4792     C101        S    D   
9     male      71      0      0  PC 17609   49.5042      NaN        C  NaN   
10    male      47      1      0  PC 17757   227.525  C62 C

# Zad2
Funkcja train_test_split dzieli losowo zbiory na dwa podzbiory: treningowy oraz testowy. Pozwala na ustalanie wielkości poszczególnych pozdbiorów.


# Zad3

In [66]:
col_name = ["cabin","CabinReduced","sex"]

# Zad4

In [67]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(data[col_name], data["survived"], test_size=0.3, random_state=0)

print(np.shape(X_train))
print(np.shape(X_test))
print(np.shape(y_train))
print(np.shape(y_test))

(916, 3)
(393, 3)
(916,)
(393,)


Zbiory zosatały podzielone zgodnie z ustalonymi parametrami,tj. zbiór testowy stanowi 30% całości zbioru.

# Zad5

In [68]:
for ZMIENNA in col_name:
    Unique_test = [x for x in X_test[ZMIENNA].unique() if x not in X_train[ZMIENNA].unique()]
    print(ZMIENNA , len(Unique_test))    

cabin 37
CabinReduced 1
sex 0


# Zad6

In [69]:
cabin = {}
for index,row in enumerate(set(data["cabin"])):
    cabin[row] = index

CabinReduced = {}
for index,row in enumerate(set(data["CabinReduced"])):
    CabinReduced[row] = index

sex = {}
for index,row in enumerate(set(data["sex"])):
    sex[row] = index   

# Zad7

In [70]:
data["cabin_map"] = data["cabin"].map(cabin)
data["CabinReduced_map"] = data["CabinReduced"].map(CabinReduced)
data["sex_map"] = data["sex"].map(sex)

In [71]:
col_name = ["cabin","cabin_map","CabinReduced","CabinReduced_map","sex","sex_map"]

X_train, X_test, y_train, y_test = train_test_split(data[col_name], data["survived"], test_size=0.3, random_state=0)

# Zad8

In [72]:
print("X_train",pd.isnull(X_train).sum(),sep='\n')
print("X_test",pd.isnull(X_test).sum(),sep='\n')
print("y_train",pd.isnull(y_train).sum(),sep='\n')
print("y_test",pd.isnull(y_test).sum(),sep='\n')

X_train
cabin               702
cabin_map             0
CabinReduced        702
CabinReduced_map      0
sex                   0
sex_map               0
dtype: int64
X_test
cabin               312
cabin_map             0
CabinReduced        312
CabinReduced_map      0
sex                   0
sex_map               0
dtype: int64
y_train
0
y_test
0


Brakujące wartości występują jedynie dla zmiennej "cabin" i analogicznie w "CabinReduced". Dla zbioru testowego jest to 30.8 % całości brakujacych wartosci, czyli niemal tyle samo co jego stosunek do całosci zbioru.

# Zad9

In [73]:
X_train = X_train.replace(np.nan,0)
X_test = X_test.replace(np.nan,0)
print("X_train",pd.isnull(X_train).sum(),sep='\n')
print("X_test",pd.isnull(X_test).sum(),sep='\n')

X_train
cabin               0
cabin_map           0
CabinReduced        0
CabinReduced_map    0
sex                 0
sex_map             0
dtype: int64
X_test
cabin               0
cabin_map           0
CabinReduced        0
CabinReduced_map    0
sex                 0
sex_map             0
dtype: int64


Podczas zamiany tych zmiennych na zmienne numeryczne, wartosci NaN zostały juz przypisane do wartosci "0". Uważam, że lepiej skorzystać z tych wartosci zamiast zamieniac w poprzednich Nan na wartosc liczbową 0 podczas gdy pozostałe wartosci są stringami.

# Zad10

In [74]:
for ZMIENNA in col_name:
    print(ZMIENNA,"test:",len(X_test[ZMIENNA].unique()),"tren:",len(X_train[ZMIENNA].unique()))

cabin test: 71 tren: 151
cabin_map test: 71 tren: 151
CabinReduced test: 8 tren: 9
CabinReduced_map test: 8 tren: 9
sex test: 2 tren: 2
sex_map test: 2 tren: 2


Unikalnosc zmiennej cabin stanowi 32% ogółu unikalnosci danych. Mapowanie nie zmieniło unikalnosci. Redukcja znacznie zmniejszyła ilość unikalnych danych i sprawiła, że jej różnica pomiwedzy zbiorami wynosi zaledwie 1.

Uważam, że proces może mieć wpływ na końcowy wynik predykcji i jakość modelu.